### DISCUSSION TITLE
Implement build config option to serve all CSS modules regardless of page served

### DISCUSSION THREAD

**Discussion by rellfy** [withastro/roadmap#909] [Proposal]
# Summary, Background & Motivation

I have been trying out Astro + HTMX and it mostly works great. Specifically, I have been using the [AHA stack](https://ahastack.dev/).

However, I have one issue which is that Astro (understandably) only serves the CSS for the initial page load. For example, if I load page `/` and then switch to page `/info` via HTMX swap, it will not have any of the `/info` styles.

A workaround can be done by importing the styles from all pages in the global stylesheet, but this is error prone and it requires global styles rather than modules. I have been trying to find a way  (maybe a build config) to have Astro serve css files by default, rather than only the ones imported by the target pages, but haven't found a way so far.

# Goals

This feature would allow for "SPA-like" apps with HTMX and css modules, allowing to apply styles to new pages the user visits without having to visit an entirely new page, since all styles are delivered on page load for all pages. By HTMX-swapping new page contents via the server, these contents would include the CSS module with the hashed class names and other identifiers, and the styling would be correctly applied.

- Add a build option to serve all CSS modules to appropriate endpoints

# Example

When fetching any web page that would normally serve an HTML with CSS modules, bundle all CSS modules from all pages rather than only from the page being requested. When new HTML is received by the client via HTMX swap, the HTML may contain the hashed CSS module class name/etc and the styles will be correctly applied by the browser.

**Comment by rellfy** [comment]
I would be happy to try implementing this feature, as it is a deal breaker for me if I can't get this functionality to work with the AHA stack. If anyone has any pointers on relevant parts of the source code I should look, I would greatly appreciate it.

**Reply by rellfy** [reply]
I created a simple integration which resolves the problem: https://github.com/rellfy/astro-merge-styles
It works for `build && preview` but not for `dev` yet.

**Reply by rellfy** [reply]
I could not find a way to make the entire functionality work with an integration, so I prepared a [PR for Astro instead](https://github.com/withastro/astro/pull/10894).
I am happy with the way it seems to work so far.

**Comment by Princesseuh** [comment]
Wouldn't using a router that support merging head works? For HTMX, it'd mean using https://htmx.org/extensions/head-support/
